### PR TITLE
Switch GithubAppInstallationController away from ja_resource and canary

### LINKED ISSUE #890: Switch GithubAppInstallationController away from ja_resource and canary
Progress on #864 

# Problem

We want to move away from the `ja_resource` and `canary` combo used in most of our controller modules. The central issue for this project is #864

There is a generic process of how to do it in that issue, as well as tracking for the milestone progress. 

The procedure is also pasted below. If this specific controller deviates from this standard procedure, feel free to ask for help here, or by opening a PR.

The PR should indicate that it's closing this issue and is progressing #864 further along.

# Procedure

- pick a controller to switch
- view actions this controller should support in `lib/code_corps_web/router.ex`
- add `action_fallback CodeCorpsWeb.FallbackController`. This makes it so in case a with clause within the controller action fails, a sort of "catch all" function will be called from `FallbackController`
- add `plug CodeCorpsWeb.Plug.DataToAttributes`. This converts the `data` hash within the conn params from the json api structure into a flat map convenient for usage with changesets
- add `plug CodeCorpsWeb.Plug.IdsToIntegers`. This casts all id parameters into integers, since this is what our authorization layer and Phoenix in general expects, while ember is sending strings.
- for each action
   - remove that action from any `load_resource`, `load_and_authorize_resource`, etc. plugs in that controller
   - remove any `def handle_#{action}` functions
   - write a `def #{action}(conn, params) do`. For the standard `create`, `update` `delete`, `show` and index actions. See examples below on how these functions should be written, usually
   - rewrite the policy function for that action to support params instead of a changeset if necessary. See section below for pointers.

That should be the whole procedure. We already have controller tests in place, so if the process has been done correctly, the tests should pass.

Some additional steps might be necessary, however, if there are some specifics to an action. For those, we will be here to provide feedback.

## How to rewrite controller actions

```Elixir
@spec index(Conn.t, map) :: Conn.t
def index(%Conn{} = conn, %{} = params) do
  with resources <- ResourceModule |> Query.id_filter(params) |> Repo.all do
    conn |> render("index.json-api", data: resources)
  end
end

@spec show(Conn.t, map) :: Conn.t
def show(%Conn{} = conn, %{"id" => id}) do
  with %ResourceModule{} = resource <- ResourceModule |> Repo.get(id) do
    conn |> render("show.json-api", data: resource)
  end
end

@spec create(Plug.Conn.t, map) :: Conn.t
def create(%Conn{} = conn, %{} = params) do
  with %User{} = current_user <- conn |> Guardian.Plug.current_resource,
       {:ok, :authorized} <- current_user |> Policy.authorize(:create, params),
       {:ok, %Comment{} = resource} <- %ResourceModule{} |> ResourceModule.create_changeset(params) |> Repo.insert do
    conn |> put_status(:created) |> render("show.json-api", data: resource)
  end
end

@spec update(Conn.t, map) :: Conn.t
def update(%Conn{} = conn, %{"id" => id} = params) do
  with %ResourceModule{} = resource <- ResourceModule |> Repo.get(id),
       %User{} = current_user <- conn |> Guardian.Plug.current_resource,
       {:ok, :authorized} <- current_user |> Policy.authorize(:update, resource),
       {:ok, %ResourceModule{} = resource} <- resource |> ResourceModule.changeset(params) |> Repo.update do
    conn |> render("show.json-api", data: resource)
  end
end
```

So to recap, the actions work like

### `index` action

- apply filters from query params to load data
- render data

### `show` action

- load resource
- render resource

### `create` action

- authorize action for current user and parameters (requires rewriting that policy to use a params map instead of a changeset, expect `{:ok, :authorized}`
- perform insert action
- render resulting resource

### `update` and `delete` actions

- load resource
- authorize action for current user on resource, expect `{:ok, :authorized}`
- perform update/delete
- render resource, or no content for delete

### Examples in code

-`CodeCorpsWeb.CommentController` has examples for standard `index`, `show`, `create` and `update` actions

## How to rewrite policies

This consists of several relatively simple steps.

- For each action being switch away, we need to move the `def can` function clause out of the `defimpl` block in `policy.ex` into the root namespace of the `CodeCorps.Policy` module. 
- We also need to make the clause private instead of public.

- If any of the clauses expect a changeset, they should now expect a simple map of params:

```Elixir
defp can?(%User{} = user, :some_action, %Changeset{data: %Resource{}} = changeset), do: Policy.Resource.some_action?(user, changeset)
```

becomes

```Elixir
defp can?(%User{} = user, :some_action, %{} = params), do: Policy.Resource.some_action?(user, params)
```

- in this case, the policy module itself also need to be rewritten to support a params map. Since in most cases, these policies simply look at the `changeset.changes` map, the rewrite ought to be simpler. If there is trouble, someone will be available to provide pointers.

### Examples in code

- the rewrite has been done for `Policy.Comment.create?`

### PR DESCRIPTION
**Pull request by treble37**
# What's in this PR?

Helps support #864

Make sure any changes to code include changes to documentation.

## References
Fixes #890

Progress on: #864

### CODE REVIEW

File: lib/code_corps_web/controllers/github_app_installation_controller.ex [hunk 1]
```diff
@@ -1,29 +1,49 @@
...
-  plug :load_resource, model: GithubAppInstallation, only: [:show], preload: @preloads
-  plug :load_and_authorize_changeset, model: GithubAppInstallation, only: [:create], preload: @preloads
-
-  plug JaResource
+  action_fallback CodeCorpsWeb.FallbackController
+  plug CodeCorpsWeb.Plug.DataToAttributes
+  plug CodeCorpsWeb.Plug.IdsToIntegers
 
   @spec model :: module
   def model, do: CodeCorps.GithubAppInstallation
```
Comment (begedin) [inline lib/code_corps_web/controllers/github_app_installation_controller.ex]: You can completely remove the `model` function. It's used internally by `ja_resource` only.

File: lib/code_corps/policy/github_app_installation.ex [hunk 2]
```diff
@@ -6,8 +6,8 @@ defmodule CodeCorps.Policy.GithubAppInstallation do
 
   alias CodeCorps.{GithubAppInstallation, User}
 
-  def create?(%User{} = user, %Ecto.Changeset{} = changeset),
-    do: changeset |> get_project |> owned_by?(user)
+  def create?(%User{} = user, params),
+    do: params |> get_project |> owned_by?(user)
 
   def update?(%User{} = user, %GithubAppInstallation{} = github_app_installation),
```
Comment (begedin) [inline lib/code_corps/policy/github_app_installation.ex]: Seems like you pointed out a bug we weren't aware of. There github app installation endpoint should only support `:index`, `:show` and `:create` in the controller. That means we do not really need an `:update` policy here, so feel free to remove it as well as the associated tests.

File: lib/code_corps_web/controllers/github_app_installation_controller.ex [hunk 3]
```diff
@@ -1,29 +1,49 @@
...
+      conn |> put_status(:created) |> render("show.json-api", data: installation)
+    end
   end
 
-  @spec handle_create(Plug.Conn.t, map) :: Ecto.Changeset.t
-  def handle_create(_conn, attributes) do
-    %GithubAppInstallation{}
-    |> GithubAppInstallation.create_changeset(attributes)
+  @spec delete(Plug.Conn.t, map) :: Conn.t
+  def delete(%Conn{} = conn, %{"id" => id} = params) do
```
Comment (begedin) [inline lib/code_corps_web/controllers/github_app_installation_controller.ex]: Again, a "bug" here. We should not be supporting a `:delete`. The controller should only have `:index`, `:show` and `:create`.

**Review by begedin — DISMISSED** [review]
Some changes that are actually unrelated to your work, but were pointed out thanks to it. Otherwise, good job!

File: lib/code_corps_web/controllers/github_app_installation_controller.ex (hunk 1 above)
Comment (treble37) [reply lib/code_corps_web/controllers/github_app_installation_controller.ex]: thx

File: lib/code_corps/policy/github_app_installation.ex (hunk 2 above)
Comment (treble37) [reply lib/code_corps/policy/github_app_installation.ex]: Ok, all changes should be live now, thanks @begedin !

File: lib/code_corps/policy/github_app_installation.ex [hunk 4]
```diff
@@ -6,9 +6,7 @@ defmodule CodeCorps.Policy.GithubAppInstallation do
 
   alias CodeCorps.{GithubAppInstallation, User}
 
-  def create?(%User{} = user, %Ecto.Changeset{} = changeset),
-    do: changeset |> get_project |> owned_by?(user)
+  def create?(%User{} = user, params),
+    do: params |> get_project |> owned_by?(user)
```
Comment (joshsmith) [inline lib/code_corps/policy/github_app_installation.ex]: Minor, but would kill this line.

**Review by joshsmith — DISMISSED** [review]

**Comment by joshsmith** [comment]
Tiny tiny change and then this is good to go.

File: lib/code_corps/policy/github_app_installation.ex (hunk 4 above)
Comment (treble37) [reply lib/code_corps/policy/github_app_installation.ex]: removed newline and it's all one one line now, thx!

**Review by begedin — APPROVED** [review]

**Comment by begedin** [comment]
Great work @treble37

**Pull request merged**